examples/DataApi-Water-And-Energy-Usage/template/modules/l0/src/main/scala/com/my/currency/l0/custom_routes/CustomRoutes.scala

```diff
@@ -50,31 +53,30 @@ case class CustomRoutes[F[_] : Async](
     ordinal: SnapshotOrdinal,
     address: Address
   ): F[AddressTransactionsWithLastRef] = {
-    context.getCurrencySnapshot(ordinal).flatMap {
-      case None => new Exception(s"Could not fetch snapshot: ${ordinal.value.value}").raiseError[F, AddressTransactionsWithLastRef]
-      case Some(value) =>
-        value.dataApplication
-          .fold(
-            AddressTransactionsWithLastRef(TxnRef.empty, List.empty[UpdateUsageTransaction]).pure
-          )(dataApplicationPart =>
-            Deserializers.deserializeState(dataApplicationPart.onChainState) match {
-              case Left(_) => new Exception(s"Could not deserialize state of snapshot: ${ordinal.value.value}").raiseError[F, AddressTransactionsWithLastRef]
-              case Right(value) =>
-                Async[F].delay {
-                  val updateUsageTransactions = value.updates.filter {
-                    _.owner == address
-                  }.sortBy(_.lastTxnOrdinal)(Ordering[SnapshotOrdinal].reverse)
-                  val referenceTransaction = updateUsageTransactions.head
+    val dataApplicationPart: F[Option[DataApplicationPart]] =
+      OptionT(context.getCurrencySnapshot(ordinal))
+        .getOrRaise(new Exception(s"Could not fetch snapshot: ${ordinal.show}, ${address.show}"))
+        .map(_.dataApplication)
+    OptionT(dataApplicationPart)
+      .semiflatMap(da => MonadThrow[F].fromEither(Deserializers.deserializeState(da.onChainState)))
+      .map(_.updates.filter(_.owner === address))
+      .map(_.sortBy(_.lastTxnOrdinal)(Ordering[SnapshotOrdinal].reverse))
+      .mapFilter(txs => txs.lastOption.map(t => AddressTransactionsWithLastRef(TxnRef(t.lastTxnOrdinal, t.lastTxnHash), txs)))
+      .getOrElse(AddressTransactionsWithLastRef(TxnRef.empty, List.empty[UpdateUsageTransaction]))
+  }
 
-                  AddressTransactionsWithLastRef(
-                    TxnRef(referenceTransaction.lastTxnOrdinal, referenceTransaction.lastTxnHash),
-                    updateUsageTransactions
-                  )
-                }
-            }
-          )
+  private def getTransactionsResponse(
+    updateUsageTransactions: List[UpdateUsageTransaction],
+    snapshotOrdinal        : SnapshotOrdinal,
+    txnHash                : String
+  ): List[TransactionResponse] =
+    updateUsageTransactions.foldLeft(List.empty[TransactionResponse]) { (acc, signedUpdate) =>
+      if (acc.isEmpty) {
+        acc :+ TransactionResponse(signedUpdate, snapshotOrdinal, txnHash)
+      } else {
+        acc :+ TransactionResponse(signedUpdate, acc.last.lastRef.txnSnapshotOrdinal, acc.last.lastRef.txnHash)
```

Comment:
FYI, this will be very slow if `updateUsageTransactions` is large because I think your algorithm is _O(n*n)_. On top of that, you're making a copy of the list for each iteration of the loop.